### ISSUE TITLE
wayland v2: don't release pressed keys without user input

### ISSUE THREAD

**Issue by lilydjwg** [fcitx/fcitx5#954]
**Describe the bug**
6f92a0cbe7a66924c552cc80d967cc32de9e929e introduced the idea to record and release pressed keys on ungrab. This caused a lot of issues when I was trying to workaround and ignore them (see e.g. https://github.com/WayfireWM/wayfire/pull/2111).

> Under sway, if a text input client exit upon a key press and the next client is XWayland, the key release will not be triggered and xwayland will produce key repitition.

This doesn't seem to happen anymore.

**To Reproduce**
1. Try to implement a Wayland compositor with global shortcuts that handles this correctly

**Expected behavior**
No superfluous key release event, so it's simpler to reason about and deal with. E.g. the issues mentioned here https://github.com/lilydjwg/wayfire/commit/f036d013d962f4f48b804f05829bbd3dea16e89f just disappear by itself.

**Desktop (please complete the following information):**
 - Desktop: Wayfire
 - Display server type: Wayland

**Additional context**

**Comment by lilydjwg** [comment]
Reproduced the issue with `wine notepad`. Will investigate.

**Comment by lilydjwg** [comment]
Could there be an option to change it?

**Comment by ammen99** [comment]
@wengxt it would be great to have an option to disable it. Right now fcitx5 has these release events as a workaround for hyprland bugs, but that breaks things on wayfire so now wayfire also has to include a workaround for the workaround that fcitx5 has for hyprland ...

**Comment by wengxt** [comment]
@ammen99 compositor should be working for whatever bad thing that client does. Another thing is all these protocol doesn't really have a clear thinking about the issues.. And also, in order to support key repetition, certain fake key release must be sent.

This is an article that describes such things.
https://www.csslayer.info/wordpress/linux/key-repetition-and-key-event-handling-issue-with-wayland-input-method-protocols/

But, come to think of it, if certain key does not repeat at all (this can be queried), actually I can make fcitx not release to it.

**Comment by ammen99** [comment]
I quickly went through your blog post and I'm wondering, why do you have to actually forward backspaces to the client (including repeat), can you not delete the surrounding text with the text-input-v3 protocol with requests from fcitx5, without the client ever knowing that there were backspaces pressed?

**Comment by wengxt** [comment]
> I quickly went through your blog post and I'm wondering, why do you have to actually forward backspaces to the client (including repeat), can you not delete the surrounding text with the text-input-v3 protocol with requests from fcitx5, without the client ever knowing that there were backspaces pressed?

@ammen99 
First of all, surrounding text support is extremely poor and buggy. I would try to avoid use it unless it is a must for certain features.
https://www.csslayer.info/wordpress/fcitx-dev/why-surrounding-text-is-the-worst-feature-in-the-linux-input-method-world/

Even if we don't consider that, certain application, specifically, terminal (alacritty, konsole, etc) can't do surrounding text. The concept of surrounding text just does not exists there. Imagine that you're using vim in a terminal and you can't do a long press backspace to delete text, that's not acceptable.

**Comment by wengxt** [comment]
Also, backspace is just one example, there are many other different keys: Like page down, page up, arrow keys those are not covered by text-input-v3. The keys that produce text can be converted to commit_string, but you know, "commit_string" and key are actually go though different codepath on the application side. Which may cause totally different behavior in application. The application is legit to implement the code as 

onKeyPressed() {
   if (key is A) {
       ...
   }
}

onTextCommit(text) {
    // Do they need to check if (text == "A") and use the same logic like onKeyPress?
    // does that even make sense? For example, the "A" can legally comes from a spell checker candidate, which is not "press A" and user won't expect it to be same as press "A".
}

**Comment by ammen99** [comment]
I see, good points, thanks for the explanation. 

The actual bug we were observing in Wayfire is actually that fcitx5 would send release keys after it is deactivated (for example user pressed alt-tab and then fcitx5 sends alt release which stops the switcher in wayfire). But I guess this also can't be avoided since deactivation could also happen for some other random reason ... This is indeed a huge mess :( I'll see what's the best we can do on Wayfire's side.

**Comment by wengxt** [comment]
@ammen99 can you also check the discussion here applies to wayfire? 

https://github.com/fcitx/fcitx5/commit/341ae16f36b2851b7dce12b89c870fd24ad44229#commitcomment-125121371

**Comment by ammen99** [comment]
> @ammen99 can you also check the discussion here applies to wayfire?
> 
> [341ae16#commitcomment-125121371](https://github.com/fcitx/fcitx5/commit/341ae16f36b2851b7dce12b89c870fd24ad44229#commitcomment-125121371)

I went through it but I'm not sure I really understood what the discussion was all about. Anyway with the modifiers, yes, if you press <alt> on the vkbd, it will remain pressed (also when switching focus) to the new app, until the alt key is released. Afaik this is how things are supposed to be.

**Comment by huyhoang8398** [comment]
did you resolve this? @lilydjwg it's kind of annoying everytime im switching app in a workspace that has terminal emulator app in wayfire

**Comment by lilydjwg** [comment]
@huyhoang8398 don't use input-method-v2. enable the "input-method-v1" plugin instead and all problems will disappear.

**Comment by huyhoang8398** [comment]
yes i specify that in the wayfire ini file to use v1, but when i have kitty or alacritty running, everytime switcher on these app, it is hanging

**Comment by lilydjwg** [comment]
@huyhoang8398 I don't have any problems switching to alacritty.
